contracts/Marketplace.sol

```diff
@@ -1,74 +1,272 @@
+    return service.versions.length;
+  }
+
+  // Manage Offer
+
+  // function getServiceOffer(string memory sid, uint offerIndex) public view returns (Offer memory offer) {
+  //   int serviceIndex = getService(sid);
+  //   Service memory service = services[serviceIndex];
+  //   // TODO: assert sid exist
+  //   // TODO: assert offer exist
+  //   return services.offers[offerIndex];
+  // }
+
+  function getServiceOffersCount(string memory sid) public view returns (uint serviceOffersCount) {
+    uint serviceIndex = getServiceIndex(sid);
+    Service memory service = services[serviceIndex];
+    return service.offers.length;
+  }
+
+  // Manage Payment
 
-    mapping (bytes32 => address) public serviceContracts;
-    bytes32[] public services;
-    uint256 public totalServices;
+  // function getServicePayment(string memory sid, uint paymentIndex) public view returns (Payment memory payment) {
+  //   uint serviceIndex = getServiceIndex(sid);
+  //   Service memory service = services[serviceIndex];
+  //   // TODO: assert sid exist
+  //   // TODO: assert payment exist
+  //   return services.payments[paymentIndex];
+  // }
+
+  function getServicePaymentsCount(string memory sid) public view returns (uint servicePaymentsCount) {
+    uint serviceIndex = getServiceIndex(sid);
+    Service memory service = services[serviceIndex];
+    return service.payments.length;
+  }
+
+  // ------------------------------------------------------
+  // Setter functions
+  // ------------------------------------------------------
+
+  // Manage Service
+
+  function createService (string memory sid) public whenNotPaused returns (uint serviceIndex) {
+    int _serviceIndex = getServiceIndexRaw(sid);
+    require(_serviceIndex == -1, "Sid is already used");
+
+    services.length++;
+    Service storage service = services[services.length - 1];
+    service.sid = sid;
+    service.owner = msg.sender;
     
-    event ServiceCreated(bytes32 sid, address serviceAddress);
-
-    constructor() Ownable() Pausable() public {}
-
-    function createService(bytes32 sid) public whenNotPaused returns (address) {
-        if (serviceContracts[sid].isContract()) {
-            return;
-        }
-        serviceContracts[sid] = new Service(msg.sender, sid);
-        services.push(sid);
-        totalServices = totalServices + 1;
-        emit ServiceCreated(sid, serviceContracts[sid]);
-        return serviceContracts[sid];
+    // The following doesn't work but seems cleaner than the previous
+    // Version[] storage versions = new Version[](0);
+    // Offer[] storage offers = new Offer[](0);
+    // Payment[] storage payments = new Payment[](0);
+    // Service storage service = Service({
+    //   sid: sid,
+    //   owner: msg.sender,
+    //   versions: versions,
+    //   offers: offers,
+    //   payments: payments
+    // });
+    // services.push(service);
+
+    emit ServiceCreated(service.owner, service.sid, services.length - 1);
```

Comment:
For what purpose do we emit ServiceCreated event?

Reply:
Every change of state functions will emit an event.
Event are the only way to create reactive application from smart contract.